contributors/devel/sig-testing/flaky-tests.md

```diff
@@ -140,34 +158,242 @@ release-blocking flakes. Therefore we have the following guidelines:
+
+Kubernetes has a lot of nested systems, so sometimes log entries can refer to
+events happening three levels deep. This means that line numbers in logs might
+not refer to where problems and messages originate. Do not make any assumptions
+about where messages are initiated!
+
+If you have trouble finding relevant logging information or events, don't be
+afraid to add debugging output to the test. For an example of this approach,
+[see this issue](https://github.com/kubernetes/kubernetes/pull/88297#issuecomment-588607417).
+
+### What should you look for?
+
+One of the first things to look for is if the test is assuming that something is
+running synchronously when it actually runs asynchronously. For example, if the
+test is kicking off a goroutine, you might need to add delays to simulate slow
+operations and reproduce issues.
+
+Examples of the types of changes you could make to try to force a failure:
+  - `time.Sleep(time.Second)` at the top of a goroutine
+  - `time.Sleep(time.Second)` at the beginning of a watch event handler
+  - `time.Sleep(time.Second)` at the end of a watch event handler
+  - `time.Sleep(time.Second)` at the beginning of a sync loop worker
+  - `time.Sleep(time.Second)` at the end of a sync loop worker
+
+Sometimes,
+[such as in this example](https://github.com/kubernetes/kubernetes/issues/93496#issuecomment-675631856),
+a test might be causing a race condition with the system it is trying to
+test. Investigate if the test is conflicting with an asynchronous background
+process. To verify the issue, simulate the test losing the race by putting a
+`time.Sleep(time.Second)` between test steps. 
+
+If a test is assuming that an operation will happen quickly, it might not be
+taking into account the configuration of a CI environment. A CI environment will
+generally be more resource-constrained and will run multiple tests in
+parallel. If it runs in less than a second locally, it could take a few seconds
+in a CI environment.
+
+Unless your test is specifically testing performance/timing, don't set tight
+timing tolerances. Use `wait.ForeverTestTimeout`, which is a reasonable stand-in
+for operations that should not take very long. This is a better approach than
+polling for 1 to 10 seconds.
+
+Is the test incorrectly assuming deterministic output? Remember that map iteration in go is
+non-deterministic. If there is a list being compiled or a set of steps are being
+performed by iterating over a map, they will not be completed in a predictable
+order. Make sure the test is able to tolerate any order in a map.
+
+Be aware that if a test is mixing random allocation with static allocation, that
+there will be intermittent conflicts.
+
+Finally, if you are using a fake client with a watcher, it can relist/rewatch at any point.
+It is better to look for specific actions in the fake client rather than
+asserting exact content of the full set.
+
+## Hunting flaky unit tests in Kubernetes
```

Comment:
I feel like we should drop this section, it looks pretty out of date (uses RC instead of RS, image last updated 5y ago, can't find source)

Reply:
I don't know if I'm enough of an expert in this material to make that call. I'll ping you about it on Slack.

Reply:
As SIG Testing chair I'm willing to make the call.  @BenTheElder @alejandrox1 WDYT?